tests/Codec/CodecLibraryTest.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+namespace MongoDB\Tests\Codec;
+
+use MongoDB\Codec\Codec;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\DecodeIfSupported;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Codec\KnowsCodecLibrary;
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Tests\TestCase;
+
+class CodecLibraryTest extends TestCase
+{
+    public function testDecode(): void
+    {
+        $codec = $this->getCodecLibrary();
+
+        $this->assertTrue($codec->canDecode('cigam'));
+        $this->assertFalse($codec->canDecode('magic'));
+
+        $this->assertSame('magic', $codec->decode('cigam'));
+    }
+
+    public function testDecodeIfSupported(): void
+    {
+        $codec = $this->getCodecLibrary();
+
+        $this->assertSame('magic', $codec->decodeIfSupported('cigam'));
+        $this->assertSame('magic', $codec->decodeIfSupported('magic'));
+    }
+
+    public function testDecodeNull(): void
+    {
+        $codec = $this->getCodecLibrary();
+
+        $this->assertFalse($codec->canDecode(null));
+
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessage('No decoder found for value of type "null"');
+
+        $this->assertNull($codec->decode(null));
+    }
+
+    public function testDecodeUnsupportedValue(): void
+    {
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessage('No decoder found for value of type "string"');
```

Comment:
Seeing this example made me realize that the default exception messages assumes codecs support is determined by _type_. In these tests, the codecs support specific string values, which makes this exception message less helpful. I don't think we need to change that but it'd be worth calling out in a comment here.

Reply:
Yes. The new exceptions will always say "Could not decode value of type <type>". The type is merely mentioned so that I don't have to worry about printing values, which may be complicated when handling arrays and such.